generator/transit_generator.cpp

```diff
@@ -370,14 +444,140 @@ bool GraphData::IsSorted() const
   return v.IsSorted();
 }
 
+void GraphData::ClipLines(vector<m2::RegionD> const & borders)
+{
+  // Set with stop ids with stops which are inside |borders|.
+  set<StopId> stopIdInside;
+  for (auto const & s : m_stops)
+  {
+    if (m2::RegionsContain(borders, s.GetPoint()))
+      stopIdInside.insert(s.GetId());
+  }
+
+  set<StopId> hasNeighborInside;
+  for (auto const & e : m_edges)
+  {
+    if (stopIdInside.count(e.GetStop1Id()) != 0 && stopIdInside.count(e.GetStop2Id()) == 0)
+      hasNeighborInside.insert(e.GetStop2Id());
+    if (stopIdInside.count(e.GetStop2Id()) != 0 && stopIdInside.count(e.GetStop1Id()) == 0)
+      hasNeighborInside.insert(e.GetStop1Id());
+  }
+
+  // Filling |lines| with stops inside |borders|.
+  vector<Line> lines;
+  for (auto const & l : m_lines)
+  {
+    Ranges const & stopIds = l.GetStopIds();
+    CHECK_EQUAL(stopIds.size(), 1, ());
+    // Note. |stopIdsToFill| will be filled with continuous sequences of stop ids.
+    // In most cases only one sequence of stop ids should be place to |stopIdsToFill|.
+    // But if a line is split by |borders| several times then several
+    // continuous sequences of stop ids will be place to |stopIdsToFill|.
+    // The loop below goes through all the stop ids belong the line |l| and keep in |stopIdsToFill|
+    // continuous sequences of stop ids which is inside |borders|.
+    Ranges stopIdsToFill;
+    bool stopWasOutsideBorders = true;
+    for (auto const stopId : stopIds[0])
```

Comment:
```c++
auto it = stopIds[0].begin();
while (it != stopIds[0].end()) {
  while (it != stopIds[0].end() && stopIdInside.count(*it) == 0)
    ++it;
  auto jt = it;
  while (jt != stopIds[0].end() && stopIdInside.count(*jt) != 0)
    ++jt;
  if (it != jt)
    stopIdsToFill.emplace_back(it, jt);
  it = jt;
}
```
No need to keep this state "stopWasOutsideBorders" and it's more declarative that you're trying to collect all consecutive groups of stops that are inside the borders.

Reply:
Yes. It's better, but code should be slightly different.
```c++
...
  auto const insideOrNeighborInside = [&](StopId stopId){
    return stopIdInside.count(stopId) || hasNeighborInside.count(stopId);
  };

  for (auto const & line : m_lines)
  {
    Ranges stopIdsToFill;
    Ranges const & ranges = line.GetStopIds();
    CHECK_EQUAL(ranges.size(), 1, ());
    vector<StopId> const & stopIds = ranges[0];
    auto it = stopIds.begin();
    while (it != stopIds.end()) {
      while (it != stopIds.end() && insideOrNeighborInside(*it) == 0)
        ++it;
      auto jt = it;
      while (jt != stopIds.end() && insideOrNeighborInside(*jt) != 0)
        ++jt;
      if (it != jt)
        stopIdsToFill.emplace_back(it, jt);
      it = jt;
    }
```